NSIS.UMUI.Complete (Suggested solution!) (Review wanted)

Tobias-B-Besemer opened the pull request:
- Moved strings into new, seperate file bleachbit_lang.nsh
- Added multi-language support for strings
- Created bleachbit_nsis_general.pot
- Created bleachbit_nsis_multiuser.pot

Tobias-B-Besemer committed:
Better multi-language support

- Moved strings into new, seperate file bleachbit_lang.nsh
- Added multi-language support for strings
- Created bleachbit_nsis_general.pot
- Created bleachbit_nsis_multiuser.pot

Tobias-B-Besemer commented:
I hope I did it right...

Lets give it a try! ;-)

az0 commented:
Did you try installation from this build https://ci.appveyor.com/project/az0/bleachbit/builds/23483634/artifacts ?

az0 commented on windows/bleachbit_nsis_general.pot:
```diff
@@ -0,0 +1,27 @@
+msgstr "Shred with BleachBit"
+
+msgid "BLEACHBIT_UPGRADE_UNINSTALL"
+msgstr "${prodname} is already installed. Click 'OK' to uninstall the old version before \ upgrading, or click 'Cancel' to abort the upgrade."
```
Does the slash \ belong in the middle here?

Tobias-B-Besemer replied:
I copied the string from you. Is it a linebreak in NSIS?

Tobias-B-Besemer replied:
OK, written in this form, it is no line-break! Also: NSIS do line-breaks by its own! So I removed the " \ "...

Tobias-B-Besemer commented:
> Did you try installation from this build https://ci.appveyor.com/project/az0/bleachbit/builds/23483634/artifacts ?

No.

Tobias-B-Besemer committed:
v2.0.0 (Debug Build)

- Included FileFunc.nsh for e.g. command line arguments managment
- Made the remaining English texts to variables
- Added icon to Windows Explorer Shell Extension
- Added descriptions to the installer options
- First command line arguments managment
- Own help page by command line
- New bleachbit_lang.nsh
- New German Translation
- New bleachbit_nsis_general.pot

Tobias-B-Besemer committed:
v2.0.1

Fixing syntax

Tobias-B-Besemer committed:
v2.0.2

Fixing syntax

Tobias-B-Besemer committed:
v2.0.3

Tobias-B-Besemer committed:
v2.0.4

Fixing + SetErrorLevel

Tobias-B-Besemer commented:
@az0, what do you mean with `FIXME later: Restore QuietUninstallString` ???
Restore QuietUninstallString if it was used before BB got de-installed because of upgrade?
Only option I see: Use of a command line parameter to pars the information to the installer...
Should we do this?

Tobias-B-Besemer commented:
Sorry, no, we can do it with `SetErrorLevel`... ^^

Tobias-B-Besemer committed:
v2.1.0

- Variables "Section" -> "Component"
- Restore QuietUninstallString
- First implementation "/no-desktop-shortcut"

Tobias-B-Besemer committed:
v2.1.1

Tobias-B-Besemer committed:
v2.1.2

Fix syntax

Tobias-B-Besemer committed:
v2.1.3

Tobias-B-Besemer committed:
v2.3.1003

Tobias-B-Besemer committed:
v2.3.1004

Tobias-B-Besemer committed:
v2.3.1005

Tobias-B-Besemer committed:
v2.3.1006

Tobias-B-Besemer committed:
v2.3.1007

Tobias-B-Besemer committed:
v2.3.1008

Tobias-B-Besemer committed:
v2.3.1009

Tobias-B-Besemer committed:
v2.3.1010

Tobias-B-Besemer commented:
OK, I doing now tests...

Tobias-B-Besemer commented:
I recognized that there exist also the bug that you can abort the uninstall of the previous version and can install anyway. I will try to fix that, too.

Tobias-B-Besemer committed:
v2.3.1011

Tobias-B-Besemer committed:
v2.3.1012

Tobias-B-Besemer commented:
I found another bug: It was always a link to the uninstaller written into the start menu - no matter if the user wanted start menu shortcuts, or not.

Tobias-B-Besemer committed:
v2.3.1013

Tobias-B-Besemer commented:
We should now have faster load times... ^^ ...and it seems it is so... ^^

Tobias-B-Besemer committed:
v2.3.1014

Tobias-B-Besemer committed:
v2.3.1015

Tobias-B-Besemer commented:
German translation works & looks wonderful! :-)

Icon for shredder in Windows Explorer works, too! :-)

Tobias-B-Besemer committed:
v2.3.1016

Tobias-B-Besemer committed:
v2.3.1017

Tobias-B-Besemer committed:
v2.3.1018

Tobias-B-Besemer committed:
v2.3.1019

Tobias-B-Besemer committed:
v2.3.1020

Tobias-B-Besemer committed:
v2.3.1021

Tobias-B-Besemer commented:
@az0, I found some bugs in the source that I have fixed, but I also implemented some features, I don't know if you will like them... So when I'm finished, just have a look, make review, and I change again what you don't like! :-)
I'm working ATM on the command line, making a /no-desktop-shortcut switch, looking that I get the bug fixed, that the installer starts anyway, even when the user abort uninstall of the old copy and my builds are not stable...

Tobias-B-Besemer committed:
v2.3.1022

Tobias-B-Besemer committed:
v2.3.1023

Tobias-B-Besemer committed:
v2.3.1024

Tobias-B-Besemer committed:
v2.3.1025

Tobias-B-Besemer committed:
v2.3.1026

Tobias-B-Besemer commented:
@az0, I improved the GUI part of the installer a bit... Even if I'm not finished with the installer, you can already have a look at it... ^^

Tobias-B-Besemer committed:
v2.3.1027

Tobias-B-Besemer committed:
v2.3.1028

Tobias-B-Besemer committed:
v2.3.1029

Tobias-B-Besemer committed:
v2.3.1030

Tobias-B-Besemer committed:
v2.3.1031

Tobias-B-Besemer committed:
v2.3.1032

Tobias-B-Besemer committed:
v2.3.1033

Tobias-B-Besemer committed:
v2.3.1034

Tobias-B-Besemer committed:
v2.3.1035

Tobias-B-Besemer committed:
v2.3.1036

Tobias-B-Besemer committed:
v2.3.1037

Tobias-B-Besemer committed:
v2.3.1038

Tobias-B-Besemer committed:
v2.3.1039

Tobias-B-Besemer committed:
v2.3.1040

az0 commented on windows/bleachbit.nsi:
```diff
@@ -17,163 +17,731 @@
+;  @tested ok v2.3.1039, Windows 7
+;  @testeddate 2019-04-03
+;  @testedby https://github.com/Tobias-B-Besemer
+;  @note 
```
I do not see a benefit to this section compared to what's already in Git. For example, Git already documents the authors and the dates of their changes. A Git commit message is a good place to add more, like how you tested it. I don't want to maintain this information.

Tobias-B-Besemer replied:
I think you don't need to maintain this information! ^^
When I'm done with the installer, I think there's nothing more to do in next time! ^^
Maybe I'm oldish, but I like those headers and even played with the idea to add a EOF like in many NSIS-examples I found...
I think people will download the source/the files, have it on there computer, work with it or spread it further... A header with version info is therefor IMHO always a good thing!
If you talk just about the test-comments: I copied the header from the cleaners... I can remove the test-stuff again, I don't need it!

az0 commented:
I appreciate your efforts here, and good results can come of it. However, this pull request is difficult for me to review, understand, and merge. The [contributor's documentation](https://github.com/bleachbit/bleachbit/blob/master/CONTRIBUTING.md) gives advice that would be a great help here. Here is the section:

>If you have multiple commits around multiple themes (such as adding two, unrelated features), please consider breaking them up into multiple pull requests by using multiple branches. Smaller pull requests are easier to review and commit.

Basically, this commit is very large, and many changes are unrelated. It's difficult to understand which line changes relate to which intention. The original intention was translations, but many changes now are related to other intentions.

Would you please split it up? When you do, it would be helpful to submit them one at a time (in series, instead of in parallel) starting with the most important change. Another strategy would be to start with the simplest change, and once that is accepted then move to more complex changes.

Other helpful advice from the contributor's document is this the following because mesages like "v2.0.3" do not carry useful information.
>See OpenStack [Git Commit Good Practice](https://wiki.openstack.org/wiki/GitCommitMessages) for advice on writing good Git commit messages. In particular, the first line of the commit message is important.

A final request relates to the many changes. I see 48 changes over 3 days. I understand it's helpful to work like this to trigger the Appveyor build, but perhaps you could make it easy to know when you think you're done and ready for review. One way is to change the title from something like "not ready to review" to "ready for review."

I'll glance over the changes and add comments.

Again, I appreciate your work, and I hope we can continue working together.

az0 commented on windows/bleachbit.nsi:
```diff
@@ -17,163 +17,731 @@
   !insertmacro MUI_LANGUAGE "Czech"
   !insertmacro MUI_LANGUAGE "Danish"
   !insertmacro MUI_LANGUAGE "Dutch"
+; !insertmacro MUI_LANGUAGE "ESPERANTO"
```
Please include translations only when there is a [BleachBit translation in Launchpad](https://translations.launchpad.net/bleachbit/trunk/+pots/bleachbit/) that is at least partially complete for two reasons

1. It increases the size of the NSIS installer
2. It ultimately does not give a good experience to the user to be able to install the software but not operate the application

What is the minimum translation? Maybe 5% or 10%. For example,  Kazakh is 2% complete, so I wouldn't provide an NSIS translation.

Tobias-B-Besemer replied:
It was just a comment and not included! Also: This comment was mainly for my future me to know which languages I can use with MUI_LANGUAGE... See my comment on top of this section in the code and my comment beyond here in the issue...

Tobias-B-Besemer replied:
Addition: I added now the languages that get provided with BleachBit already!
The users should IMHO be able to install BB in this language, they also use in BB.
So "Farsi" is e.g. "Persian" and "PortugueseBR" is "Basilian Portuguese"...

az0 commented on windows/bleachbit.nsi:
```diff
@@ -17,163 +17,731 @@
+    Delete "$SMSTARTUP\BleachBit.lnk"
+
+    ; Remove file association (Shredder)
+    DeleteRegKey HKCR "AllFileSystemObjects\shell\shred.bleachbit"
```
I see a huge amount of code changes, and I'm not sure what the purpose of all them is. Please see my comments about splitting up the PR and commit messages

Tobias-B-Besemer replied:
I also resorted the code (the functions) for better (logical) readability... But GitHub don't "understand" this... So e.g. this lines are still (unchanged) code from you! It have just a other position in the document...

Tobias-B-Besemer replied:
Addition: If you have now a look at my files at a whole, it should have saved use many time!
I re-wrote many things and re-structured them, so you have to watch the changes now just ones...
...but I will in future make better commit messages! I promise! ^^
But not yet, to many changes, I add them on the bottom (with screenshots they show) as comment... ^^

az0 commented on windows/bleachbit_lang.nsh:
```diff
@@ -0,0 +1,1810 @@
+!endif
+
+!ifdef LANG_GERMAN
+	; Translation 2019-04-01 - 2019-04-02 by https://github.com/Tobias-B-Besemer
```
It seems this whole file should be generated by a script from the `.po` files downloaded from Launchpad. Managing the translation another way could cost much effort, lead to copying errors, and cause delays.

Here's the program I use to do something like this
https://github.com/bleachbit/bleachbit-misc/blob/master/extract_desktop.py

az0 replied:
If you want my help with this, let me know. I might be able to make fast progress if I had a table or map that translates the ISO codes used for .po files to NSIS Language names. For example, de=German, es=Spanish

Anyway, the the goal would be to let translators work in Launchpad, and then the script makes it easy to get the translations from Launchpad into this NSIS file

Tobias-B-Besemer replied:
This should IMHO be done in a new PR!
I opened a issue for it (#529) and wrote something to it...
As long will BB only have complete support for German... But this is IMHO not so terrible as long we get the from me generated PO files into Launchpad. Because the translators need time to translate the strings...

Tobias-B-Besemer replied:
Addition: In my latest release I have done a lot about language! Also set the ISO codes used for .po files in relation to NSIS Language names... But later more to it!

az0 commented on windows/bleachbit_lang.nsh:
```diff
@@ -0,0 +1,1810 @@
+	LangString BLEACHBIT_COMPONENT_AUTOSTART_TITLE ${LANG_GERMAN} "Autostart"
+	LangString BLEACHBIT_COMPONENT_AUTOSTART_DESCRIPTION ${LANG_GERMAN} "Erzeugt Autostart-Verknüpfung für BleachBit"
+	LangString BLEACHBIT_COMPONENT_TRANSLATIONS_TITLE ${LANG_GERMAN} "Übersetzungen"
+	LangString BLEACHBIT_COMPONENT_TRANSLATIONS_DESCRIPTION ${LANG_GERMAN} "Ble
```
Like my other comment, let's please include only the languages in the NSIS installer for which the BleachBit application has decent translations.

Tobias-B-Besemer replied:
I got equal with NsisMultiUserLang.nsh. But if you let the NSH files be auto-generated, I will change it again. But in the next PR! I will still need that data for #529 and the other languages are now just comments...

Tobias-B-Besemer replied:
Addition: That's my German translation you pointed to! ^^
Language Strings not used in MUI_Language doesn't get compiled!

az0 commented on windows/bleachbit_nsis_multiuser.pot:
```diff
@@ -0,0 +1,93 @@
+# English translation of bleachbit_nsis_multiuser.pot
+# Copyright (C) 2019 Andrew Ziem
```
This file looks like OK, but to work with Launchpad, it needs to be integrated into bleachbit.pot using in `po/Makefile`. Look at the Makefile to see how it combines `cleanerml.pot` with `desktop.pot`. If you need help, I can do it.

Tobias-B-Besemer replied:
OK, I also add this comment to issue #529...

Tobias-B-Besemer committed:
v2.3.1041

az0 reviewed (changes requested):
Please see several comments I added in two places: the GitHub conversation (general comments) and some comments attached to specific lines of code.

Tobias-B-Besemer commented:
Just in short... Read you msg later again... But it is no more possible to break this thing in peaces! Sorry! I worked on the no-desktop-icon-switch-story and the multi-user-addon of NSIS made it hard with the commend line parameters... You see in my pushes: Many fails! I will be happy when it's stable and making peaces again out of it seem not possible for me... Sorry! It's a complete new version. But good documented! Read the complete file! No diff! I have made - like ever - many comments... ;-)

Tobias-B-Besemer committed:
v2.3.1043

Tobias-B-Besemer commented:
I write later more to your feedback...

Tobias-B-Besemer commented:
> A final request relates to the many changes. I see 48 changes over 3 days. I understand it's helpful to work like this to trigger the Appveyor build, but perhaps you could make it easy to know when you think you're done and ready for review. One way is to change the title from something like "not ready to review" to "ready for review."

OK, I misused the Appveyor a bit... ;-) :-D ...but I really should have debug builds... :-/
...but I think I found now a better way. I write you in issue https://github.com/bleachbit/wishlist/issues/38 about it...
And I changed now the title to "NSIS (Status: Debugging)"...

> Other helpful advice from the contributor's document is this the following because mesages like "v2.0.3" do not carry useful information.
>> See OpenStack Git Commit Good Practice for advice on writing good Git commit messages. In particular, the first line of the commit message is important.

I use commit messages like "v2.0.3" if I find no better one, if it's a debugging build and when the changes are so many that it is a new version. ;-)
But I guess you wanted to know my changes here for review and for the readme...
Btw.: I didn't added them by now, because (as you can now read) they are not finished/tested and the builds are not stable/don't get created.

But here they are:
+ I found many bugs in the source;
+ Resorted the code (read above about it) and made it more pretty; (a lot of changes in diff come from this)
+ Language support with NsisMultiUserLang.nsh & bleachbit_lang.nsh + POTs; (TESTED !!!)
+ Multi-Language lizense support; (not finished)
+ "Current User" should use for shortcuts the users folders and the program should be registered/use in the registry in HKCU - for "All Users" the installer should write the shortcuts in shared folders and should register/use in the registry HKLM. Therefor also added Macro/Function "SetShellVarContext"; (not finished)
+ Adding the command line switch "/no-desktop-shortcut" and keep the switches from "NSIS Multi User Plugin" working; (not tested)
+ To keep function "/uninstall" (after adding "/no-desktop-shortcut") made Macro/Function "Uninstall"; (not tested)
+ Added: Command line "/uninstall" should use internal uninstaller; (not tested) (almost the same as above)
+ Added: Error level 666 for QuietUninstallString + restore; (not tested)
+ Fixed the "uninstaller canceled -> installer continuous upgrade anyway"-bug; (not tested)
+ Fixed uninstaller should be used from the installation; (not tested)
+ MUI_RESERVEFILE_LANGDLL for faster startup; (works!)
+ Played with some MUI options; (most work)
+ Added the start and ending pages with pictures to the installer & uninstaller; (need maybe compression/better art-work)
+ MUI_FUNCTION_DESCRIPTION for components; (works!)
+ Added to the uninstaller components -> set registry key for modify, then make that the user can add/remove/repair translations/shortcuts/shredder; (not finished)
+ Uninstaller use now lang strings MUI_FUNCTION_DESCRIPTION -> changed them; (finished!)
+ "Start automatically" -> "Autostart" (was because of space, but describes also better); (finished)
+ "Integrate Shred" -> "Shredder for Explorer" (because of use of string in uninstaller); (finished)
+ A lot of small changes;
+ Added a lot of documentation to the code;
+ Maybe some things I forgot... ;-) :-D

To-Dos:
+ Uninstaller should know install size -> read them from registry;
+ Installer should set date installed in registry;
+ Installer/uninstaller components horizontal for preparation for packages;
+ Error level 665 for SystemComponent + restore

...and that are just the changes I done on the installer in the last ~5 days... ^^
If you really want to write all in the readme what I have done the lat ~4 months, the readme will get long... ;-)
...so I suggest to just write "improved ..., ..., ..., cleaners" & "improved the installer/uninstaller".

OK, I push now a new build... I write you later more.

Tobias-B-Besemer committed:
Debugging

Tobias-B-Besemer commented:
OK, I have it completely broken... ^^
> AppVeyor was unable to build non-mergeable pull req...

...and...
> Conflicting files
windows/NsisInclude/NsisMultiUserLang.nsh

Tobias-B-Besemer commented:
OK, finally it's time for some feedback...

> I appreciate your efforts here, and good results can come of it. However, this pull request is difficult for me to review, understand, and merge. The [contributor's documentation](https://github.com/bleachbit/bleachbit/blob/master/CONTRIBUTING.md) gives advice that would be a great help here. Here is the section:
> 
> > If you have multiple commits around multiple themes (such as adding two, unrelated features), please consider breaking them up into multiple pull requests by using multiple branches. Smaller pull requests are easier to review and commit.
> 
> Basically, this commit is very large, and many changes are unrelated. It's difficult to understand which line changes relate to which intention. The original intention was translations, but many changes now are related to other intentions.
> 
> Would you please split it up? When you do, it would be helpful to submit them one at a time (in series, instead of in parallel) starting with the most important change. Another strategy would be to start with the simplest change, and once that is accepted then move to more complex changes.

I thought long about what writing back, because it's a difficult topic and English is not my first language... But OK, just give it a try... ^^
The thing is, that my intension was to push the BleachBit Project a bit forward and so I spend the last ~4 months all my spare-time for this project and have now ~50 PR open...
...but this "high energy"-phase was time-limited because I deferred everything else (all my other projects)...
Doesn't mean that I'm go - it means just that I will in future/next time don't spend so much time anymore for the project.
My plans are to finish and do some things (sub-projects) in the next ~2 weeks and after that, still be there to change the things you request in reviews and to land the PRs, but mainly spend my time with other projects, again...
...after some time, when I have done a lot of other things for the other things in my life, I will "come back" and spend more time again (but later more to it...).
So this brings me to the point about splitting this PR into very small parts...
You should see this installer as a complete new one, that just includes some lines of your old code!
Splitting this into small parts and submitting them part by part will take me months!
So I beg you to review this PR as a whole one!
Just FYI, but I think you already know it: You can take Notepad++ for this.
Notepad++ have code highlighting for NSIS-files, too and is with it much much better then GitHub.
Also: I suggest you to read the complete files, not just the diffs, because it makes IMHO more sense complete then in peaces. (And you can take your old source as reference...)
OK, time for dinner, later more... ^^

Tobias-B-Besemer commented:
OK, lets come to the next topic...
I spend the last 12 days with working on this installer/uninstaller... and I will need some more, if I will really perfect finish them...
So here's my problem:
I have already ~50 PRs open and you be maybe busy IRL and can't do review of them...
...but I spend already a lot of time for it and I don't wanna waste my time with something you never check in!
So if you promise me, that "when you change this issues ... then I will check it in!", then I will spend some more days to finish it, but I'm ATM on the point I fear that you never will check it in... :-/

Tobias-B-Besemer commented:
OK, now to my changes I'm done in the last days... ^^

But first: I normally planed that for BBv3.0, but if GTK+3.0 lands in BBv2.4, then maybe a switch to the "Ultra Modern UI" in BBv2.4 would also be cool!
You can have a look at it here: http://ultramodernui.sourceforge.net/
Till now I have done work on Command Line and functionality in general... GUI can be split from it... Maybe in a new PR... But if you like the "Ultra Modern UI", I can write a PR for it.
I have also other ideas about what's also possible with the NSIS System, but later more to it...

So, and if we are by the GUI, then lets start with the GUI of the installer/uninstaller... ^^

This is how the BB Installer looks like now...

First comes always a language dialog, that the user (e.g. Win10 Multi-Language) can always change the installer language... Selected Language get saved and suggested next time you use the installer...
![2019-04-12 15_20_29-Installer Language](https://user-images.githubusercontent.com/11881073/56040525-df733f80-5d36-11e9-9b53-7d29a4e99734.jpg)

...then follows the startpage...
![2019-04-12 15_20_59-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56040556-ec902e80-5d36-11e9-9f25-0e941afd2263.jpg)

...as written before, art-work maybe needs to get perfected and the pics re-sampled for smaller size...
...if we do a switch to "Ultra Modern UI" in BBv2.3, the pics are obsolete and we don't have to waste time working on them.

After that comes the license dialog...
![2019-04-12 15_21_13-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56040583-fd40a480-5d36-11e9-944d-7ea7dd8a24fb.jpg)

Use must accept it to proceed!
I included now GPLv3s in any language I found! Not really many! But anyway...
The files got converted by me from HTML or PDF to TXT... Therefor the Beta Tester should have a look on them about formatting, special characters, etc....
Arabic is ATM included as PDF and not shown in MUI-Licensedialog - conversion with Adobe Reader and/or C&P didn't worked with the Arabic special characters! I need to find a other way, later. ...

Then the Multi-User-Selection and the Directory-Selection as before, but now with logo...
![2019-04-12 15_34_02-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56041203-78568a80-5d38-11e9-973b-7361550a14f8.jpg)
![2019-04-12 15_34_11-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56041212-7ab8e480-5d38-11e9-8df2-86aea5c1db5e.jpg)

The Components Page have now a horizontal design for Install Types (the groups)...

...and I found a new bug! The sections text is on my system always in German, even if I start the installer in English, and I can't find the bug and have no clue! ^^

![2019-04-12 18_10_35-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56051253-510ab800-5d4e-11e9-9b22-6e0794f121ef.jpg)

I created three Install Types: Full, Typical, Minimal
(Can't make a screenshot - the list collapse when I'm try!)

"Cleaners" are now a collapsed Section Group.
In a normal build only "Stable" is shown and selected.
"Alpha" gets activated with "!ifdef AlphaTester" and includes "BetaTester".
"Beta" gets activated with "!ifdef BetaTester".
The path to CleanerML is defined with "!define PathToCleanerML" on top of bleachbit.nsi.

"Installer" is the installer that can be now copied to the destination path.
The installer have now Add/Remove (/Repair) Functionality !!! ^^
But later more to it...

"Translations" is now also a Section Group!
With on section per language!
That we don't need translation and the user don't need to speak English, the section text to the languages are now written in the language that gets installed !!!

"Minimal" installation have now the same features as the "English only"-Installer!

And now the clue: "Typical" installs just "English (USA)" (build in in BleachBit) and the language the user installs with !!! ^^
(And in my case German, but that's the bug mentioned above...)

![2019-04-12 18_24_27-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56051862-3df8e780-5d50-11e9-8155-06dd5fb4c205.jpg)

Another bug that's make me mad today: I get an error by coping the installer to the destination path!
Maybe that's my Avira... This software s*ckZ! But I can't uninstall it ATM, because I have to many (important) files in quarantine and the f... uninstaller of Avira kills all by uninstall! So: Need to investiagte that later...

![2019-04-12 15_36_42-Installation von BleachBit](https://user-images.githubusercontent.com/11881073/56052157-02aae880-5d51-11e9-8222-b7a4c882b24a.jpg)

And the finish page:

![2019-04-12 18_30_30-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56052211-1e15f380-5d51-11e9-893c-353e66ec91da.jpg)

Tobias-B-Besemer commented:
OK, now to the "Add/Remove" Functionality:

The installer have now "Add/Remove (/Repair)" Functionality !!! ^^

Therefor some things changed...
+ The install check the installed version when he gets started
+ Same Version: Add/Remove Functionality
+ Older Version: De-install
+ De-install is now silent to make sure all components get de-installed and not just some
+ If installed version is newer, the installer stops with a error message
+ There's a command line switch "/downgrade" to overwrite the version check ^^
+ Installed Components get saved in registry under HKCU/Software/BleachBit/Components,
or under HKLM/Software/BleachBit/Components
+ Uninstall process by upgrade gets started with the command line switch /upgrade
+ /upgrade don't deletes the components list -> So the new version gets installed with the same components

To-Dos:
+ BleachBit should also recognize Cleaners installed in sub-folders -> Then the installer can better install & uninstall Alpha & Beta Cleaners

Tobias-B-Besemer commented:
About the EXE-Files:
The EXE-Files have now information in the Details Tab!
But Windows requires Version in format 1.2.3.4!
So you should change your versioning to e.g. 2.3.0.1049, or 2.4.0.0.

![2019-04-12 18_43_30-Eigenschaften von BleachBit-2 3 0 1049-Setup-Alpha-Tester exe](https://user-images.githubusercontent.com/11881073/56053007-2707c480-5d53-11e9-85e9-a52f56186789.jpg)

Tobias-B-Besemer commented:
About the NSIS-Files:
+ The NSIS-Files are now split in multiple files!
+ One file is a LVC (LV-Crew, my crew) Addon for NsisMultiUser that provides additionally Command Line Functionality
+ One file is a BB (BleachBit) Addon for NsisMultiUser that provides additionally GUI Functionality like the Add/Remove Feature
+ The other files are language files

Tobias-B-Besemer commented:
The uninstaller:
+ Almost the same as the installer;
+ Only installed components gets selected;
+ Estimated Size of Install gets now displayed.

But:
+ The Estimated Size of Install is ATM hard-coded
+ Install Date is still not saved in registry
+ QuietUninstallString is a ADDITION to UninstallString -> Admins should use SystemComponent -> Must change a lot of code
+ "installer.exe /curentuser /uninstall" works, but "installer.exe /allusers /uninstall" give me this German Error Message and I have no clue again what's going on! ^^ Need to debug this...
![2019-04-12 15_36_57-C](https://user-images.githubusercontent.com/11881073/56053586-ced1c200-5d54-11e9-94fd-bcc1015daaa9.jpg)

OK, I hope I have nothing forgotten... ^^

Tobias-B-Besemer commented:
At the end some screenshots from the uninstaller... ^^

![2019-04-12 19_00_29-Installer Language](https://user-images.githubusercontent.com/11881073/56053922-b8783600-5d55-11e9-9114-2bea05c5c2bd.jpg)
![2019-04-12 19_00_39-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56053923-b8783600-5d55-11e9-87aa-b6212ad8d66c.jpg)
![2019-04-12 19_00_45-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56053924-b8783600-5d55-11e9-99f3-9fdf5add9fac.jpg)
![2019-04-12 19_02_48-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56053926-b8783600-5d55-11e9-9d21-e9bd34db4b1b.jpg)
![2019-04-12 19_03_02-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56053927-b910cc80-5d55-11e9-9efe-f683405472f5.jpg)
![2019-04-12 19_03_11-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56053928-b910cc80-5d55-11e9-8ecf-ae1655ce3749.jpg)

Tobias-B-Besemer commented:
I try now to solve the conflicts in this PR and make a new push with the latest version.

Beta Testing (of functionality) would be nice!

Review is not needed now, but also welcome!

Btw.: The POT-Files are outdated !!!

Tobias-B-Besemer committed:
Merge remote-tracking branch 'upstream/master' into NSIS

Tobias-B-Besemer committed:
See PR #527

Tobias-B-Besemer committed:
Just a guess to get my 4 digits for appveyor

Tobias-B-Besemer committed:
v2.3.0.1052

Tobias-B-Besemer commented:
@az0, bit complicated to get here a build... But NSIS works @ home...

az0 commented:
I appreciate your hard work, but I don't know where to begin. The complexity of this is intimidating.

Tobias-B-Besemer commented:
At the beginning! ^^
As I said: See it as a complete new installer! ^^
If you don't do review because it's intimidating, I can stop working on it... :-/

az0 commented:
OK, how about this: Would you please write: one concise, up-to-date list of all the changes and another list of all the outstanding problems or concerns. (Currently it's spread over multiple comments over multiple days, and it's hard to tell whether one comment supersedes another.) I will review at this level.

Don't do it yet, but would you be willing to write a test plan I could give to another developer to check it in depth?

Tobias-B-Besemer commented:
OK, I worked now several hours on it and this is what came out by it:
I would write a test plan, but fast I figured out, that a questions list ("Does ... work?") for Alpha/Beta/Nightly Testers is/should be much better.
The file is now called "bleachbit.nsi_-_tests.txt", is located in the windows folder and comes with my next push.

I also thought about landing this patch and had the following idea:
Just do a short review of it, not in deep, only to look that I don't delete the complete PC with my installer. ^^
Then land this patch - this installer is (with all his remaining bugs) already much better then the current one! ^^
After landed the patch and having the installer in the Nightlies, inform the Alpha/Beta/Nightly Testers about it - in the forum and/or on the homepage.
Then give the first part of my questions (GUI Installer/Uninstaller, Multi-Language-Support, Add/Remove Functionality, System Integration - No CLI testing!) to them that I/we get a feedback about bugs.
In the meantime I will fix the issues I already know.
Now, after knowing you will try to land this PR, I'm willing to spend e.g. one more week on fixing/improving it. (In peaces - what no more really possible is, because the solutions are now nested - it would have taken a year, and I don't have the time for it...)
After all, when I'm "finished" you can make a review "in deep" - this should also save you much time, because you don't have to always review again my changes on my own code. ^^
I think it would be really nice, if we get this landed/stable till v2.4.0.xxxx.

K, also had a short look on the "Ultra-Modern UI"...
Was easier as I thought - I just had to modify 3 lines of code. ^^
In the following I will post you screenshots from how the "standard install [skin]" of UMUI looks ATM in the installer/uninstaller.
You now need to decide if you want to have it!
UMUI brings a lot of new, cool & good features we can use!
But you make the decision what UI the installer/uninstaller should have!

Oh, almost forgot it:
The next pull also includes a files called "bleachbit.nsi_-_to-dos.txt", in the windows folder.
But the To-Dos-List only includes things I can't do alone!
A much bigger list of open bugs/tests/etc. is on my PC - but the list is written in "Keywords", not intended to leaf my working place! ^^

And the "Features List":
Do you really need a detailed list ???
Main changes are Bug Fixes, Add/Remove Functionality, switch "/no-desktop-shortcut", QuietUninstallString/SystemComponent & general improvements.
There is a help menu on CLI and the GUI is self-explaining.
If you want more details, maybe read the list of questions to the testers - I'll try to test every changed/improved/fixed feature! ^^
But if you really need a detail list, I'll write one! I just don't know how detailed it should be...

K, screenshots + push later... ^^

Tobias-B-Besemer commented:
OK, here the screenshots of the installer with UMUI and the default skin... (Yes, UMUI among other things skin-able! ^^)
![2019-04-13 01_52_11-Installer Language](https://user-images.githubusercontent.com/11881073/56076652-0e8bbe80-5dd4-11e9-86e2-91a807be5ad3.jpg)
![2019-04-13 01_52_33-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076654-0e8bbe80-5dd4-11e9-827e-ef693a83c51f.jpg)
![2019-04-13 01_52_51-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076655-0e8bbe80-5dd4-11e9-8e63-25f96618665c.jpg)
![2019-04-13 01_53_09-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076656-0f245500-5dd4-11e9-83cc-2b27e22cc019.jpg)
![2019-04-13 01_53_20-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076657-0f245500-5dd4-11e9-8b6d-a51f9a7962e8.jpg)
![2019-04-13 03_23_22-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076658-0f245500-5dd4-11e9-9b33-fe0df358cca4.jpg)
![2019-04-13 03_24_10-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076659-0f245500-5dd4-11e9-885d-4510ce58e22b.jpg)

Tobias-B-Besemer commented:
And here the screenshots of the uninstaller...
![2019-04-13 03_24_54-Installer Language](https://user-images.githubusercontent.com/11881073/56076682-409d2080-5dd4-11e9-8249-9b9d81ea1361.jpg)
![2019-04-13 03_25_02-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56076683-409d2080-5dd4-11e9-89ce-362c1daf0cc3.jpg)
![2019-04-13 03_25_11-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56076684-4135b700-5dd4-11e9-9568-470b7e99ecd6.jpg)
![2019-04-13 03_25_28-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56076685-4135b700-5dd4-11e9-9ea8-918b2d326c05.jpg)
![2019-04-13 03_25_40-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56076686-4135b700-5dd4-11e9-89d7-e24d9782e2d3.jpg)
![2019-04-13 03_25_51-BleachBit Uninstall](https://user-images.githubusercontent.com/11881073/56076687-4135b700-5dd4-11e9-87ae-0adcc8973e34.jpg)

Tobias-B-Besemer commented:
And I did already - with my restricted GFX Skills - a replacement of the left side blurry PC... ^^
But only one screenshot from it! It looks on every page the same! ^^

![2019-04-13 06_09_49-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56076738-b2756a00-5dd4-11e9-8304-953a5e54fac8.jpg)

If you make your decision to pro UMUI - and as long we have no GFX Artist for it - you can also make your decision about the background color I should use! Black, white, blue, which color it should be ??? ^^

Tobias-B-Besemer commented:
OK, this is my concept for the new installer... ^^

![2019-04-13 11_15_43-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56077578-3122d500-5dde-11e9-95e7-5faeadfc08e6.jpg)

Tobias-B-Besemer committed:
Mainly UMUI (+ more)

Tobias-B-Besemer committed:
Try to build the builds

Tobias-B-Besemer commented:
Andrew, if you decide to use UMUI (what I suggest you), then you should update NSIS on AppVeyor with a build from here:
https://sourceforge.net/projects/ultramodernui/files/UltraModernUI/UltraModernUI%202.0b5/
Else we don't get AppVeyor to build the NSIS File that include the UMUI Code...

Tobias-B-Besemer committed:
Mainly cleaning up

Tobias-B-Besemer commented:
I recognized ATM that UMUI have own functionality to add the Add/Remove Feature! :-O
Means: All the time I spend and work I done for it gets obsolete and I can remove it again! ;-(
Benefit: The code gets in future managed/updated by someone else who is really better then me! ^^
And you can be happy that you didn't done review already - it's all away again... ;-) :-D
I just need your OK now to use UMUI! ^^

Tobias-B-Besemer committed:
Just a sync...

az0 commented on .gitignore:
```diff
@@ -15,6 +15,7 @@ locale/
 
 # Windows build
 build/
+cleanerml/
```
This looks like it doesn't belong

Tobias-B-Besemer replied:
This is for making automated BetaTester & AlphaTester Builds (that include the cleaners from "release" & "pending" in installation, as own topic/sub-section). I can program in NSIS the direct path to the CleanerML Repo, but unfortunately I can relative just go one path level higher and not two. (Can just use "..\cleanerml" and not "...\cleanerml".)

az0 commented:
>Just do a short review of it, not in deep, only to look that I don't delete the complete PC with my installer. ^^ Then land this patch - this installer is (with all his remaining bugs) already much better then the current one! 

At this time I would accept it into its own branch

>I think it would be really nice, if we get this landed/stable till v2.4.0.xxxx.

I am skipping to version 3.0. The reason is it is becoming urgent to change to GTK 3 and Python 3. The first change is GTK 3, and there is a branch for that. These could require changes to the NSIS installer, and I am hoping to avoid merge conflicts.

>But the To-Dos-List only includes things I can't do alone!

If they are not done, will there be regressions against version 2.0 or signficiant new bugs? 

>Main changes are Bug Fixes, Add/Remove Functionality, switch "/no-desktop-shortcut", QuietUninstallString/SystemComponent & general improvements.

The "bug fixes" and "general improvements" are not clear enough to either test now or to explain in the release notes for the next final release. If the change is tiny (such as a cosmetic change), then it does not need to be documented.

If end users would notice an enhancement or bug fix, then it's worth document for the release notes.

If a change could cause a regression or maybe would not work equally well on all systems, then I'd prefer to note it for testing.

>Benefit: The code gets in future managed/updated by someone else who is really better then me! ^^

In general, I consider this a great benefit. I am a proponent of the DRY programming principle. 

>Andrew, if you decide to use UMUI (what I suggest you), then you should update NSIS on AppVeyor with a build from here: https://sourceforge.net/projects/ultramodernui/files/UltraModernUI/UltraModernUI%202.0b5/
>Else we don't get AppVeyor to build the NSIS File that include the UMUI Code...

This means the screenshots are from your personal system because UMUI does not currently work on AppVeyor, correct?

I prefer to use the mainline NSIS. Looking at the Git repositories for UMUI and the mainline NSIS, it seems UMUI maybe is out of date. Also, it's not clear it will stay up to date.

az0 commented:
Did you fix this [silent install](https://www.bleachbit.org/forum/silent-install-doesnt-seem-work#comment-3192) bug? I'd like him to test your version, but there hasn't been a successful build on AppVeyor for 13 days

Tobias-B-Besemer commented:
> > Just do a short review of it, not in deep, only to look that I don't delete the complete PC with my installer. ^^ Then land this patch - this installer is (with all his remaining bugs) already much better then the current one!
> 
> At this time I would accept it into its own branch

That's great!

> > I think it would be really nice, if we get this landed/stable till v2.4.0.xxxx.
> 
> I am skipping to version 3.0. The reason is it is becoming urgent to change to GTK 3 and Python 3. The first change is GTK 3, and there is a branch for that. These could require changes to the NSIS installer, and I am hoping to avoid merge conflicts.

My plan is to be finished with the NSIS Installer in the next 2-3 weeks. If I integrate them in the Master Branch, or a GTK 3 Branch doesn't matter much for me... NSIS change nothing on the BleachBit Python Code and the GTK 3 Branch gets integrated into the Master Branch again... My plan was to have a new, fresh, funky, professional installer for GTK 3, that the new GUI gets also reflected by the installer...

> > But the To-Dos-List only includes things I can't do alone!
> 
> If they are not done, will there be regressions against version 2.0 or signficiant new bugs?

No! Mainly it's get AppVeyor to build the installers again...
You can have a look into "bleachbit.nsi_-_to-dos.txt" that is now pushed with this PR....

> > Main changes are Bug Fixes, Add/Remove Functionality, switch "/no-desktop-shortcut", QuietUninstallString/SystemComponent & general improvements.
> 
> The "bug fixes" and "general improvements" are not clear enough to either test now or to explain in the release notes for the next final release. If the change is tiny (such as a cosmetic change), then it does not need to be documented.

You can bring it together with "better 'current user'/'all users' system integration".

> If end users would notice an enhancement or bug fix, then it's worth document for the release notes.
> 
> If a change could cause a regression or maybe would not work equally well on all systems, then I'd prefer to note it for testing.

Please have a look into "bleachbit.nsi_-_tests.txt" that includes details questions/tests to Beta-/Alpha-Tester about the changes in NSIS.

> > Benefit: The code gets in future managed/updated by someone else who is really better then me! ^^
> 
> In general, I consider this a great benefit. I am a proponent of the DRY programming principle.
> 
> > Andrew, if you decide to use UMUI (what I suggest you), then you should update NSIS on AppVeyor with a build from here: https://sourceforge.net/projects/ultramodernui/files/UltraModernUI/UltraModernUI%202.0b5/
> > Else we don't get AppVeyor to build the NSIS File that include the UMUI Code...
> 
> This means the screenshots are from your personal system because UMUI does not currently work on AppVeyor, correct?

Yes! I already improved the design a bit more, but yes, that's what I have compiled!

> I prefer to use the mainline NSIS. Looking at the Git repositories for UMUI and the mainline NSIS, it seems UMUI maybe is out of date. Also, it's not clear it will stay up to date.

Last update to UMUI was January 2019! GitHub is just a new copy/place for the source! Version 1.00 beta 1 was released March 20, 2005! I'm pretty sure the project will persist! ^^
Here is the official home page: http://ultramodernui.sourceforge.net/
Main things seems to be done (like for NSIS in general, too) on sourceforge.net...
If you trust my evaluation and go with me the UMUI way, I can remove a lot of my code (that related to the Add/Remove Feature), bring a new professional skin to the installer and additional features...
...if you stay on MUI2, you will have more to review and still a oldish interface and I must revert my changes belonging to UMUI... :-/

Tobias-B-Besemer commented:
> Did you fix this [silent install](https://www.bleachbit.org/forum/silent-install-doesnt-seem-work#comment-3192) bug? I'd like him to test your version, but there hasn't been a successful build on AppVeyor for 13 days

Yes, should be fixed in my builds! ^^ But I write you more to it in the other issue...
I think I will upload my EXEs to Google Drive for testing... ;-)
That there are no more successful builds on AppVeyor comes because you need to change some things on AppVeyor that my builds get generated again... :-/

Tobias-B-Besemer commented:
With this Link you should get access to my BleachBit Project Files:
https://drive.google.com/drive/folders/1ZbzJimB2rPPaC9zen-5_8frIX5RlCN4M?usp=sharing

Tobias-B-Besemer commented:
OK, new builds with MUI2 & UMUI are online! ;-)

You can give it a try!

Tobias-B-Besemer commented:
Andrew, I reverted now the UMUI changes on the snapshot I had and "finished" the MUI2 builds...

Parallel I mad a other branch on my local disc and worked (and "finished") the UMUI builds... ^^

So there are now two kind of builds: MUI2 & UMUI ^^

In the following I will post here some screenshots from my latest UMUI builds to convince you ^^ that UMUI is the better solution... ^^

...then I'm going to push here the latest UMUI sources...

Parallel I'm opening a new PR with the MUI2 source...

I have already a PR "NSIS.Merge (Review wanted)" open...
There I try to go with you the first steps together.
(MUI2 & UMUI build have many in common and I will push small peaces for you... ^^)

Btw.: "My work is so far done!" and I should have testers and feedback about bugs and suggestions to improve the builds more...

Tobias-B-Besemer commented:
OK, one really nice thing beside the ability to skin the installer/uninstaller and that IMHO UMUI looks more like a professional solution... ^^ (...and BleachBit reserves that!) ...is, that you have all in one Window! The language dialog, Add/Remove, even the build-in Updater! ^^
![2019-04-17 15_26_11-Installation von BleachBit](https://user-images.githubusercontent.com/11881073/56365244-aa606480-61f0-11e9-9c8c-07bf9a754bfe.jpg)

Here is a screenshot of the build-in Updater with my/our custom warning:
![2019-04-18 07_27_11-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56365333-e09de400-61f0-11e9-90a6-e09af8b79746.jpg)

UMUI have a "nice" Install Type selection and a "menu" for ADD/Remove that looks the same:
(The modify page have also the warning, again.)
![2019-04-17 15_27_18-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56365487-47bb9880-61f1-11e9-9e55-c8581f4b7ed2.jpg)
![2019-04-18 05_08_50-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56365503-4db17980-61f1-11e9-8705-dc95470fa4d2.jpg)

I have even renamed a bit the Section Text that the "Confirm Installation" Dialog looks good... ^^
![2019-04-18 07_28_06-BleachBit Setup](https://user-images.githubusercontent.com/11881073/56365703-cadcee80-61f1-11e9-86fa-9488c9ab03aa.jpg)

Give it a try! My newest builds are here for download: (Also MUI2 builds...)
https://drive.google.com/drive/folders/1ZbzJimB2rPPaC9zen-5_8frIX5RlCN4M?usp=sharing

Tobias-B-Besemer committed:
v2.3.0.1059 - 2019-04-18

Tobias-B-Besemer committed:
GTK3 Support / Makefiles.bat

+ GTK3 Support
+ Makefiles.bat
+ Autostart: Show minimized

Tobias-B-Besemer commented:
Latest builds:
https://drive.google.com/drive/u/0/folders/1ZbzJimB2rPPaC9zen-5_8frIX5RlCN4M

Tobias-B-Besemer committed:
Gray Skin

Tobias-B-Besemer commented:
New builds are available:
https://drive.google.com/drive/u/0/folders/1ZbzJimB2rPPaC9zen-5_8frIX5RlCN4M

az0 commented:
@Tobias-B-Besemer 
Is this pull request ready for review? I am arranging for someone to review it in depth.

az0 commented:
@Tobias-B-Besemer 

There is this error from AppVeyor
```
Processing script file: "windows\bleachbit.nsi" (ACP)
2019-04-24 20:13:26 - ERROR - !include: could not find: "NsisIncludeOthers\German.nsh"
Error in script "windows\bleachbit.nsi" on line 258 -- aborting creation process
2019-04-24 20:13:26 - ERROR - windows\BleachBit-2.3.0.0-Setup.exe not found
Command exited with code 1
```

Also, there are many merge conflicts.

What do you think of me holding off the review until you can review these issues?

Tobias-B-Besemer commented:
Generally the installer is ready for review.
Can't remember the old errors, but they should be gone.
Have to make a new push with the BB3.0 changes. Removed the 3 languages you removed, but haven't pushed no more...
The source isn't lint yet, but I think we can lint it later, 'cause it compiles and work as it is...
You should first switch to the 3, or 4 dot versioning, 'cause AppVeyor will not finish else. After that I must maybe a push again, to get our changes (my suggested ones and then your really done) sync...
Would suggest you to start review with the UI. Download last version and have a look on feel & handle and give me feedback to it...

Tobias-B-Besemer commented:
Oh, and you have to install the UMUI installer for this PR!
UMUI installer is basically the same as the normal installer but have included some plugins.
Zhe UMUI installer is able to compile but versions...
As you know, I highly recommend the UMUI installer because of look, feel and simpler code by same functionality...
...and the project is also ~10yo and persists...

Tobias-B-Besemer commented:
Just for the notes, 3 things open to be done:
- Lint code
- https://github.com/SuperPat45/UltraModernUI/issues/12
- https://github.com/SuperPat45/UltraModernUI/issues/13

Tobias-B-Besemer commented:
The following PR will include the following changes:
- Remove the 3 removed languages and update the NSIS files to the latest version;
- Update files in NsisInclude - this should fix some of the conflicts;
- Update the forgotten files in NsisPluginsUnicode;
- Remove "XPStyle on" as noted in https://github.com/SuperPat45/UltraModernUI/issues/12 and add a comment for future to it.

Tobias-B-Besemer committed:
Bring lates changes to PR

Changes: https://github.com/bleachbit/bleachbit/pull/527#issuecomment-552477571

Tobias-B-Besemer committed:
Lint version

Tobias-B-Besemer committed:
"LegalCopyright" = "Andrew Ziem"

Tobias-B-Besemer committed:
Remove the "XPStyle On"

az0 commented:
I apologize in my delays in responding to this PR. We now do have a working system to localize NSIS, which is integrated with Weblate, so that overlaps significantly with the intent of this PR. Also, there are major merge conflicts (again---apologizes).
